compressors/shannon_coder.py

```diff
@@ -0,0 +1,156 @@
+"""Shannon Coding
+Shannon coding is a basic prefix free code. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses cumulative probability based method for shannon coding as described in the wiki article above.
+"""
+from typing import Any, Tuple
+from utils.bitarray_utils import float_to_bitarrays, BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeEncoder, PrefixFreeDecoder, PrefixFreeTree, BinaryNode
+from core.prob_dist import ProbabilityDist
+import math
+
+
+class ShannonTree(PrefixFreeTree):
+    """
+    Generates Codewords based on Shannon Coding. Allows generating PrefixFreeTree from the obtained codewords.
+    """
+
+    def __init__(self, prob_dist: ProbabilityDist):
+        self.prob_dist = prob_dist
+        # sort the probability distribution in decreasing probability and get cumulative probability which will be
+        # used for encoding
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        self.cum_prob_dict = self.sorted_prob_dist.cumulative_prob_dict
+        # construct the tree and set the root_node of PrefixFreeTree base class
+        super().__init__(root_node=self.build_shannon_tree())
+
+    def encode_alphabet(self, symbol) -> BitArray:
+        # compute the mid-point corresponding to the range of the given symbol
+        cum_prob = self.cum_prob_dict[symbol]
+
+        # compute encode length
+        encode_len = math.ceil(self.prob_dist.log_probability(symbol))
+
+        # the encode is the binary representation of the cumulative probability in ascending order,
+        _, code = float_to_bitarrays(cum_prob, encode_len)
+        return code
+
+    @staticmethod
+    def _build_tree_from_code(symbol, code, root_node) -> BinaryNode:
+        """ function to generate prefix-free tree from code.
+        Args:
+            symbol: current symbol
+            code: current code
+            root_node: root node to the ShannonTree
+
+        Returns:
+            the pointer to root node of the tree so far
+        """
+        # initialize the curr_node, code_so_far temporary var
+        curr_node = root_node
+        code_so_far = BitArray()
+        code_len = len(code)
+
+        for i, bit in enumerate(code):
+            # if pointer is None, the object is not updated
```

Comment:
add 
```
i.e., curr_node.right_child being None would not allow us to keep the right child in a variable for updating further.
```

Reply:
On further thought, if we are having this trouble again and again, maybe we want to make the __init__ behavior of a node to instantiate the left and right child?

Reply:
I have updated the comment but in general I like your suggestion. @kedartatwawadi: thoughts?

Reply:
Raised issue #43